### PR TITLE
Add basic initial developer docs for TSDB

### PR DESCRIPTION
**Pull request by Dieterbe**
There's a decent amount of content already out there (blog posts,
conference talks, etc), but:
* when they get stale, they don't tend to get updated
* they still leave me with questions that I'ld like to answer
  for developers (like me) who want to use, or work with, TSDB

What I propose is developer docs inside the prometheus
repository.  Easy to find and harness the power of the community
to expand it and keep it up to date.

* perfect is the enemy of good.  Let's have a base and incrementally improve
* Markdown docs should be broad but not too deep.  Source code comments
  can complement them, and are the ideal place for implementation details.

Signed-off-by: Dieter Plaetinck <dieter@grafana.com>

### CODE REVIEW

**Comment by beorn7** [comment]
I leave the details to more @codesome reviewers… but perhaps it would make sense to explain the problem of Go Modules picking up the semantic versioning tag of the prometheus/prometheus repo and wrongly interpreting it as a versioning of Prometheus as a library? I.e. mention the workaround with pinning to a certain commit, something like https://github.com/prometheus/prometheus/issues/7663#issuecomment-663672223 (I'm actually not sure what Cortex maintainers do these days to work around the issue.)

**Comment by roidelapluie** [comment]
Yes this is a major annoyance. One of the big issue for developers is that the godoc website is completely out of date, we should mention it as well.

**Comment by LeviHarrison** [comment]
I've found [Go Walker](https://gowalker.org) to be a nice solution for this. It generates the documentation directly from main, not tied to a Go Modules version. For example: https://gowalker.org/github.com/prometheus/prometheus/tsdb.

File: tsdb/docs/usage.md
Comment (codesome) [reply tsdb/docs/usage.md:39]: Yes

File: tsdb/head.go
Comment (codesome) [inline tsdb/head.go:1463]: That's not true. It refers to the ChunkID for mmappedChunks[0], and mmappedChunks[1] being firstChunkID+1, and so on. With firstChunkID+len(mmappedChunks) being the head chunk.

File: tsdb/head.go
Comment (codesome) [inline tsdb/head.go:1466]: Right about the sentence starting `Probably...` (but would not keep it in the code)
```suggestion
	// We keep the last 4 samples here (in addition to appending them to the chunk) so we don't
	// need coordination between appender and querier for the chunk bytes that are still being written to.
	sampleBuf     [4]sample
```

File: tsdb/head.go
Comment (codesome) [inline tsdb/head.go:1469]: It is nil only if headChunk is nil. And that is possible if there was an appender that created a new series, but rolled back the commit (the first sample would create a headChunk, hence appender, but rollback skipped it while the Append() call would create a series).
```suggestion
	app chunkenc.Appender // Current appender for the head chunk. Set when a new head chunk is cut.
```

File: tsdb/head_append.go
Comment (codesome) [inline tsdb/head_append.go:117]:
```suggestion
	appendID, cleanupAppendIDsBelow := h.iso.newAppendID() // Every appender gets an ID that is cleared upon commit/rollback.
```

File: tsdb/head_append.go
Comment (codesome) [inline tsdb/head_append.go:398]:
```suggestion
// Commit writes to the WAL and adds the data to the Head.
```

File: tsdb/head_append.go
Comment (codesome) [inline tsdb/head_append.go:346]:
```suggestion
// log writes all headAppender's data to the WAL.
```

File: tsdb/head_append.go
Comment (codesome) [inline tsdb/head_append.go:570]: > removes the headAppender's data from the series

This bit is not correct, it only discards the data from the appender (because we have not added to series yet at this point, it is only the isolation handling that you see below) while WAL logging the series records if any.
```suggestion
// Rollback removes the samples and exemplars from headAppender and writes the series to WAL if any.
```

File: tsdb/head_read.go
Comment (codesome) [inline tsdb/head_read.go:178]:
```suggestion
// chunkID returns the ID corresponding to .mmappedChunks[pos] (head chunk if pos==len(mmappedChunks)).
```

File: tsdb/head_read.go
Comment (codesome) [inline tsdb/head_read.go:355]: That's correct
```suggestion
// iterator returns a chunk iterator for the requested chunkID.
```

File: tsdb/head_read.go
Comment (codesome) [inline tsdb/head_read.go:459]:
```suggestion
// memSafeIterator returns values from the wrapped stopIterator
```

File: tsdb/head_read.go
Comment (codesome) [inline tsdb/head_read.go:505]: It is indeed after stopAfter. `i` is initialised to -1 above. A newStopIterator() can make it more easily understandable in case you want to add, but out of scope of this PR .
```suggestion
// stopIterator wraps an Iterator, but only returns the first
// stopAfter values.
```

File: tsdb/head.go
Comment (Dieterbe) [reply tsdb/head.go:1463]: I think what you're saying is a superset of what i'm saying?
What's wrong with my comment? - can you make a suggestion?

File: tsdb/head.go
Comment (Dieterbe) [reply tsdb/head.go:1466]: > Right about the sentence starting Probably... (but would not keep it in the code)

I cleaned it up, but i think it's super useful to keep this in the code (without the "probably"), the question of "why do we keep exactly 4 samples in a separate buffer" was one of the great mysteries of TSDB to me, and the insight that we need to cover one byte and that 4 samples is what it takes, even in the most compact encoding case, is IMHO a big revelation that should be documented.

File: tsdb/head_read.go
Comment (Dieterbe) [reply tsdb/head_read.go:505]: ah i see. it honors stopAfter if initialized with i=-1 (i assumed it would be initialized with the default zero value). I'll add a note for this.

File: tsdb/chunks/chunks.go
Comment (bwplotka) [inline tsdb/chunks/chunks.go:71]:
```suggestion
// Iterator iterates over the chunks of a single time series.
```

File: tsdb/chunks/head_chunks.go
Comment (bwplotka) [inline tsdb/chunks/head_chunks.go:137]:
```suggestion
	maxt      int64 // Max timestamp among all of this file's chunks.
```

File: tsdb/docs/usage.md
Comment (bwplotka) [inline tsdb/docs/usage.md:75]: This will age really badly. Why we cannot refer to tests? Or at least put it as `func Example()` in unit test suite?

File: tsdb/docs/usage.md
Comment (bwplotka) [inline tsdb/docs/usage.md:107]: ditto

**Review by bwplotka — CHANGES_REQUESTED** [review]
Amazing work. I would love to port those example to unit test of `func Example..`  to have it buildable and up to date. WDYT?

See https://go.dev/blog/examples

**Comment by Dieterbe** [comment]
> I would love to port those example to unit test of func Example.. to have it buildable and up to date. WDYT?

@bwplotka I think it's a great idea, but as @roidelapluie said, godoc public docs for prometheus are currently out of date / for the wrong version.

Is there a link that I can/should use to point to the most current godoc for tsdb? Or can somebody fix the doc versioning issue? (it's out of scope for me).
Once this is resolved, i will be happy to move those code examples from the markdown into the code.

File: tsdb/head.go
Comment (codesome) [reply tsdb/head.go:1478]: Not really. ChunkID referes to the id that corresponds to `mmappedChunks[0]`, and `firstChunkID+i` for `mmappedChunks[i]`, `mmappedChunks[0].ref` is the chunk reference used to access m-mapped chunk from the disk (the 4/4B packed ref).

File: tsdb/head.go
Comment (codesome) [reply tsdb/head.go]: The latest update makes sense, thanks

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:11]:
```suggestion
Callers should use [`tsdb.Open`](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#Open) to open a TSDB
(the directory may be new or pre-existing).
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:14]:
```suggestion
A `DB` has the following main components:
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:17]: There are also persistent blocks. You could point it to https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#DB.Blocks.

Also, what was the reason to point leveled compactor it to the codebase and not https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#LeveledCompactor?

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:26]:
```suggestion
* [WAL](https://pkg.go.dev/github.com/prometheus/prometheus/tsdb/wal#WAL) (Write Ahead Log).
* [`stripeSeries`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/head.go#L1279):
  this holds all the active series by linking to [`memSeries`](https://github.com/prometheus/prometheus/blob/1270b87970baeb926fcce64552db5c744ffaf83f/tsdb/head.go#L1449)
  by an ID (aka "ref") and by labels hash.
* Postings list (reverse index): For any label-value pair, holds all the corresponding series refs. Used for queries.
* Tombstones.
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:32]: uff, as pointed out previously, the golang docs have the wrong signature for the Appender. This could confuse developers more than it helps. Maybe we should fix the docs _before_ merging this PR (or make this and other godoc links point replaced with the codebase instead, pinned to the latest version and mention the version in this file).

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:38]:
```suggestion
* Use `Commit()` to add the samples to the DB and update the WAL.
* Create a new appender each time you commit.
* Appenders are not concurrency safe, but scrapes run concurrently and as such, leverage multiple appenders concurrently.
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:46]:
```suggestion
  * `tsdb.min-block-duration/2` older than the max time in the Head block. Note that while technically `storage.tsdb.min-block-duration` is configurable, it's a hidden option and changing it is discouraged.  So We can assume this value to be 2h.
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:47]: A newline above, else it is being rendered with the above point.
```suggestion

  Breaching this condition results in "out of bounds" errors.  
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:51]:
```suggestion
3) If the sample, for the respective series (based on all the labels) is out of order or has a different value for the last (highest) timestamp seen. (results in `storage.ErrOutOfOrderSample` and `storage.ErrDuplicateSampleForTimestamp` respectively)
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:53]:
```suggestion
`Commit()` may also refuse data that is out of order with respect to samples that were added via a different appender.
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:87]: Added comments and minor edits to existing comments.

```suggestion
	series := labels.FromStrings("foo", "bar")
	// ref is 0 for the first append since we don't know the reference for the series.
	ref, err := app.Append(0, series, time.Now().Unix(), 123)
	time.Sleep(time.Second)
	// Re-using the ref from above since it's the same series, makes append faster.
	_, err = app.Append(ref, series, time.Now().Unix(), 124)
	err = app.Commit()
	noErr(err)
	app = db.Appender(ctx) // Need a new one. See app.Commit() docs.
```

File: tsdb/docs/usage.md
Comment (codesome) [inline tsdb/docs/usage.md:128]: Why context.TODO() here and context.Background() for the ingestion?

**Review by codesome — COMMENTED** [review]
Generally happy with usage.md, most of my comments are minor nits.

One thing though: the godocs are outdated, and some interfaces have changed, so pointing to it here can confuse developers more than it helps. And also what @bwplotka mentioned about the code ageing badly.

But I don't thing either of us have time to fix and maintain the godoc updates (at least not right now? or is it simple to include in the CI pipeline?). So here is my suggestion to unblock this PR (while we try to fix godoc in parallel):
1. Pin the docs to the latest version by mentioning that this docs was last updated for v2.xx.x.
2. Replace all the current links (godoc and github) to point to the corresponding line of code for the latest comment in the respective release branch (example, for release `v2.31.x`, you will find a branch named `release-2.31`). PS: need not update links for new patch releases, they contain only bugfixes.

@bwplotka WDYT about this temporary solution?

File: tsdb/docs/usage.md
Comment (Dieterbe) [reply tsdb/docs/usage.md:17]: > Also, what was the reason to point leveled compactor it to the codebase and not https://pkg.go.dev/github.com/prometheus/prometheus/tsdb#LeveledCompactor?

no good reason.  until we decide on how to address the godoc problem, i have them a bit intermixed.

**Comment by codesome** [comment]
@Dieterbe let's go ahead with my suggestion in https://github.com/prometheus/prometheus/pull/9451#pullrequestreview-800390671 for now - mentioning v2.31.1 in the usage.md and replace all links (current github and godoc) to point to code lines in https://github.com/prometheus/prometheus/tree/411021ada9ab41095923b8d2df9365b632fd40c3

**Comment by codesome** [comment]
Just found out with @Dieterbe that we can use the commit for godoc
Here is the one for v2.31.1: https://pkg.go.dev/github.com/prometheus/prometheus@v1.8.2-0.20211105201321-411021ada9ab

**Comment by Dieterbe** [comment]
I have addressed all outstanding items. In particular pinned the docs/links and moved the example code to Example functions as part of the unit tests.

File: tsdb/example_test.go
Comment (codesome) [inline tsdb/example_test.go:1]:
```suggestion
// Copyright 2021 The Prometheus Authors
```

File: README.md
Comment (codesome) [inline README.md:114]:
```suggestion
  * The source code is periodically indexed, but due to an issue with versioning, the "latest" docs shown on Godoc are outdated. Instead, you can use [the docs for v2.31.1](https://pkg.go.dev/github.com/prometheus/prometheus@v1.8.2-0.20211105201321-411021ada9ab).
```

**Review by codesome — APPROVED** [review]
LGTM with nits! We should be linking example_test.go in the usage.md for it to be discoverable.

File: tsdb/example_test.go
Comment (codesome) [inline tsdb/example_test.go:33]: Since this is in _test.go, we can have it like a proper unit test, without panics, but with assertions

**Comment by codesome** [comment]
(Fixed them myself, will merge on green)

**Pull request merged**
